core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,300 @@
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.underfs.SeekableUnderFileInputStream;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.IdUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalListeners;
+import com.google.common.cache.RemovalNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An under filesystem input stream manager that can cache seekable input streams for future reuse.
+ * The manager caches the input streams, with a time-based eviction policy of expiration configured
+ * in {@link PropertyKey.WORKER_UFS_INSTREAM_CACHE_EXPIRE_MS}.
+ *
+ * An under storage file maybe opened with multiple input streams at the same time. The manager uses
+ * {@link UfsInputStreamIds} to track the in-use input stream and the available ones. The manager
+ * closes the input streams after they are expired and not in-use anymore.
+ */
+@ThreadSafe
+public class UfsInputStreamManager {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsInputStreamManager.class);
+
+  /**
+   * A map from the ufs file name to the metadata of the input streams. Synchronization is needed to
+   * access this map.
+   */
+  private final HashMap<String, UfsInputStreamIds> mFileToInputStreamIds;
+  /**
+   * Cache of the input streams, from the input stream id to the input stream.
+   */
+  @GuardedBy("mFileToInputStreamIds")
+  private final Cache<Long, SeekableUnderFileInputStream> mUnderFileInputStreamCache;
+  /**
+   * Thread pool for asynchronously removing the expired input streams.
+   */
+  private final ExecutorService mRemovalThreadPool;
+
+  /**
+   * Creates a new {@link UfsInputStreamManager}.
+   */
+  public UfsInputStreamManager() {
+    mFileToInputStreamIds = new HashMap<>();
+    mRemovalThreadPool = Executors.newFixedThreadPool(2);
+
+    // A listener to the input stream removal.
+    RemovalListener<Long, SeekableUnderFileInputStream> listener =
+        (RemovalNotification<Long, SeekableUnderFileInputStream> removal) -> {
+          SeekableUnderFileInputStream inputStream = removal.getValue();
+          synchronized (mFileToInputStreamIds) {
+            if (mFileToInputStreamIds.containsKey(inputStream.getFilePath())) {
+              UfsInputStreamIds resources = mFileToInputStreamIds.get(inputStream.getFilePath());
```

Comment:
Is it possible for someone else to hold an active reference to `resources` while not holding the `mFileToInputStreamIds` lock?

Reply:
All the methods hold `mFileToInputStreamIds` lock before access `resources` except for `acquire`. `acquire` releases the `mFileToInputStreamIds` lock after it obtains the `resources`. This is an optimization given the knowledge that the subsequent operation won't modify `mFileToInputStreamIds `.